x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/action/ShardFollowNodeTaskTests.java

```diff
@@ -5,205 +5,623 @@
+
+    private Queue<Exception> readFailures;
+    private Queue<Exception> writeFailures;
+    private Queue<Exception> mappingUpdateFailures;
+    private Queue<Long> imdVersions;
+    private Queue<Long> followerGlobalCheckpoints;
+
+    public void testCoordinateReads() {
+        ShardFollowNodeTask task = createShardFollowTask(8, 8, 8, Integer.MAX_VALUE, Long.MAX_VALUE);
+        startTask(task, 64, -1);
+
+        task.coordinateReads();
+        assertThat(shardChangesRequests.size(), equalTo(8));
+        assertThat(shardChangesRequests.get(0)[0], equalTo(0L));
+        assertThat(shardChangesRequests.get(0)[1], equalTo(8L));
+        assertThat(shardChangesRequests.get(1)[0], equalTo(9L));
+        assertThat(shardChangesRequests.get(1)[1], equalTo(8L));
+        assertThat(shardChangesRequests.get(2)[0], equalTo(18L));
+        assertThat(shardChangesRequests.get(2)[1], equalTo(8L));
+        assertThat(shardChangesRequests.get(3)[0], equalTo(27L));
+        assertThat(shardChangesRequests.get(3)[1], equalTo(8L));
+        assertThat(shardChangesRequests.get(4)[0], equalTo(36L));
+        assertThat(shardChangesRequests.get(4)[1], equalTo(8L));
+        assertThat(shardChangesRequests.get(5)[0], equalTo(45L));
+        assertThat(shardChangesRequests.get(5)[1], equalTo(8L));
+        assertThat(shardChangesRequests.get(6)[0], equalTo(54L));
+        assertThat(shardChangesRequests.get(6)[1], equalTo(8L));
+        assertThat(shardChangesRequests.get(7)[0], equalTo(63L));
+        assertThat(shardChangesRequests.get(7)[1], equalTo(8L));
+
+        ShardFollowNodeTask.Status status = task.getStatus();
+        assertThat(status.getNumberOfConcurrentReads(), equalTo(8));
+        assertThat(status.getLastRequestedSeqno(), equalTo(64L));
+    }
+
+    public void testCoordinateReads_writeBuffer() {
+        // Need to set concurrentWrites to 0, other the write buffer gets flushed immediately:
+        ShardFollowNodeTask task = createShardFollowTask(64, 1, 0, 32, Long.MAX_VALUE);
+        startTask(task, 64, -1);
+
+        task.coordinateReads();
+        assertThat(shardChangesRequests.size(), equalTo(1));
+        assertThat(shardChangesRequests.get(0)[0], equalTo(0L));
+        assertThat(shardChangesRequests.get(0)[1], equalTo(64L));
+
+        shardChangesRequests.clear();
+        // Also invokes the coordinatesReads() method:
+        task.innerHandleReadResponse(0L, 63L, generateShardChangesResponse(0, 63, 0L, 128L));
+        assertThat(shardChangesRequests.size(), equalTo(0)); // no more reads, because write buffer is full
+
+        ShardFollowNodeTask.Status status = task.getStatus();
+        assertThat(status.getNumberOfConcurrentReads(), equalTo(0));
+        assertThat(status.getNumberOfConcurrentWrites(), equalTo(0));
+        assertThat(status.getLastRequestedSeqno(), equalTo(64L));
+        assertThat(status.getLeaderGlobalCheckpoint(), equalTo(128L));
+    }
+
+    public void testCoordinateReads_maxConcurrentReads() {
```

Comment:
Can we avoid using camel_case here? I think `testMaxConcurrentReads` should be good.